### PR TITLE
leaky_relu forward-over-reverse rule

### PR DESCRIPTION
**Pull request by zou3519**
Derivation in comments

Test Plan:
- run tests

Fixes #ISSUE_NUMBER

### CODE REVIEW

File: tools/autograd/derivatives.yaml [hunk 1]
```diff
@@ -2216,6 +2216,18 @@
...
+  # so the jvp formula is the following:
+  # grad_output_t * at::where(self_p > 0, self_p.new_ones([]), negative_slope);
+  #
+  # leaky_relu_backward(grad_output, result, negative_slope, true)
+  # computes grad_output * at::where(result > 0, 1, negative_slope)
+  # so the jvp formula is the following:
+  # grad_output_t * at::where(result_p > 0, result_p.new_ones([]), negative_slope);
+  #
+  # Combined together that results in the following optimized kernel:
+  result: leaky_relu_backward(grad_output_t, self_p, negative_slope, self_is_result)
```
Comment (soulitzer) [inline tools/autograd/derivatives.yaml:2230]: Any idea what self_is_result does here? It looks like were are computing the same thing.

**Review by soulitzer — APPROVED** [review]
Thanks! Just have one small question.

File: tools/autograd/derivatives.yaml (hunk 1 above)
Comment (zou3519) [reply tools/autograd/derivatives.yaml:2230]: Ah good question. 

If `self_is_result` is True, and `negative_slope` is negative, then we have a problem. There's a check for that in https://github.com/pytorch/pytorch/blob/10023eeb65d7bc6c71f49bf0ce7eaf5b58d52fd7/aten/src/ATen/native/Activation.cpp#L121-L126 already: if the self_is_result is True and negative_slope is negative, then it raises an error.

I will beef up the comment to mention this

**Pull request closed without merging**